tools/harness-simulation/posix/sniffer_sim/sniffer.py

```diff
@@ -28,104 +28,172 @@
 #
 
 import argparse
+from concurrent import futures
+import enum
+import grpc
+import ipaddress
 import logging
 import signal
-import time
 import pcap_codec
-import sys
 import threading
 
+from proto import sniffer_pb2
+from proto import sniffer_pb2_grpc
 import sniffer_transport
 
 
-class Sniffer:
-    """ Class representing the Sniffing node, whose main task is listening.
-    """
+class SnifferServicer(sniffer_pb2_grpc.Sniffer):
+    """ Class representing the Sniffing node, whose main task is listening. """
 
-    logger = logging.getLogger('sniffer.Sniffer')
+    logger = logging.getLogger('sniffer.SnifferServicer')
 
     RECV_BUFFER_SIZE = 4096
+    MAX_NODES_NUM = 33
 
-    def __init__(self, filename, channel):
-        self._pcap = pcap_codec.PcapCodec(filename, channel)
-
-        # Create transport
-        transport_factory = sniffer_transport.SnifferTransportFactory()
-        self._transport = transport_factory.create_transport()
+    class State(enum.Enum):
+        STOPPED = 0
+        RUNNING = 1
 
+    def _clear(self):
+        self._state = SnifferServicer.State.STOPPED
+        self._pcap = None
+        self._allowed_nodeids = None
+        self._transport = None
         self._thread = None
-        self._thread_alive = threading.Event()
         self._thread_alive.clear()
 
+    def __init__(self):
+        self._thread_alive = threading.Event()
+        self._mutex = threading.Lock()  # for self._allowed_nodeids
+        self._clear()
+
     def _sniffer_main_loop(self):
         """ Sniffer main loop. """
 
         self.logger.debug('Sniffer started.')
 
         while self._thread_alive.is_set():
+            # Avoid being blocked endlessly when there is no data
+            if not self._transport.ready(0.1):
+                continue
             data, nodeid = self._transport.recv(self.RECV_BUFFER_SIZE)
-            self._pcap.append(data)
+
+            self._mutex.acquire()
+            allowed_nodeids = self._allowed_nodeids
+            self._mutex.release()
```

Comment:
```suggestion
            with self._mutex:
                sniffed = id in self._allowed_nodeids
```

nit: more efficient without copying the whole nodeids

Reply:
I think `allowed_nodeids = self._allowed_nodeids` will not copy the whole list. It just makes `allowed_nodeids` point to an existing list. The time complexity of the assignment is constant regardless of the size of the list.

Reply:
Actually, according to the Python Global Interpreter Lock (GIL), the multiple threads will never run on different CPU cores simultaneously, in which case the atomic assignment does not require a mutex. However, I think it is better to keep the mutex for readability.

Reply:
I see. We are using the lock to protect the assignment of self. _allowed_nodeids.
SGTM.